keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1330 @@
+
+List of cases where the current solution is not optimal:
+
+1. Without extra manual effort, an application running with a single replica has to settle for
+   experiencing application downtime during the node drain. They cannot use PDBs with
+   `minAvailable: 1` or `maxUnavailable: 0`, or they will block node maintenance. Not every user
+   needs high availability either, due to a preference for a simpler deployment model, lack of
+   application support for HA, or to minimize compute costs. Also, any automated solution needs
+   to edit the PDB to account for the additional pod that needs to be spun to move the workload
+   from one node to another. This has been discussed in issue [kubernetes/kubernetes#66811](https://github.com/kubernetes/kubernetes/issues/66811)
+   and in issue [kubernetes/kubernetes#114877](https://github.com/kubernetes/kubernetes/issues/114877).
+2. Similar to the first point, it is difficult to use PDBs for applications that can have a variable
+   number of pods; for example applications with a configured horizontal pod autoscaler (HPA). These
+   applications cannot be disrupted during a low load when they have only pod. However, it is
+   possible to disrupt the pods during a high load without experiencing application downtime. If
+   the minimum number of pods is 1, PDBs cannot be used without blocking the node drain. This has
+   been discussed in issue [kubernetes/kubernetes#93476](https://github.com/kubernetes/kubernetes/issues/93476)
+3. Graceful deletion of DaemonSet pods is currently only supported as part of (Linux) graceful node
+   shutdown. The length of the shutdown is again not application specific and is set cluster-wide
+   (optionally by priority) by the cluster admin. This does not take into account 
+   `.spec.terminationGracePeriodSeconds` of each pod and may cause premature termination of
+   the application. This has been discussed in issue [kubernetes/kubernetes#75482](https://github.com/kubernetes/kubernetes/issues/75482)
+   and in issue [kubernetes-sigs/cluster-api#6158](https://github.com/kubernetes-sigs/cluster-api/issues/6158).
+4. There are cases during a node shutdown, when data corruption can occur due to premature node
+   shutdown. It would be great if applications could perform data migration and synchronization of 
+   cached writes to the underlying storage before the pod deletion occurs. This is not easy to
+   quantify even with pod's `.spec.shutdownGracePeriod`, as the time depends on the size of the data
+   and the speed of the storage. This has been discussed in issue [kubernetes/kubernetes#116618](https://github.com/kubernetes/kubernetes/issues/116618)
+   and in issue [kubernetes/kubernetes#115148](https://github.com/kubernetes/kubernetes/issues/115148).
+5. There is not enough metadata about why the node drain was requested. This has been discussed in
+   issue [kubernetes/kubernetes#30586](https://github.com/kubernetes/kubernetes/issues/30586).
```

Comment:
NMO includes a [reason](https://github.com/medik8s/node-maintenance-operator/blob/main/api/v1beta1/nodemaintenance_types.go#L51-L53) for the maintenance procedure.

Furthermore, [NMO's status](https://github.com/medik8s/node-maintenance-operator/blob/main/api/v1beta1/nodemaintenance_types.go#L57-L85) adds some in-progress and post-progress information on the maintenance status. The operator also [emits events](https://github.com/medik8s/node-maintenance-operator/blob/main/pkg/utils/events.go) about the maintenance process.